src/mqtt_client.rs

```diff
@@ -132,9 +142,10 @@ where
         Ok(read)
     }
 
-    fn write(&self, buf: &[u8]) -> Result<(), Error<N::Error>> {
+    fn write(&self, state: &mut SessionState<C>, buf: &[u8]) -> Result<(), Error<N::Error>> {
```

Comment:
This is a pretty message way of handling ownership - let's just use the session state inside of `self` and address any ownership issues resulting from that at their original source.

Reply:
I'd argue this commit doesn't add to the mess; the mess already exists in the current version, with state split between the interior-mutable SessionState and the regular rest. I really don't want to refactor that in this PR.